File: packages/cli/src/templates/cc-native/_cc-native/lib/state.py
"""
CC-Native State Management Module.

Handles plan state file operations and iteration tracking for the review process.

State files are stored adjacent to plan files (e.g., ~/.claude/plans/foo.state.json)
to prevent state loss when session IDs change or temp files are cleaned up.
"""

import json
import sys
from datetime import datetime
from pathlib import Path
from typing import Any, Dict, Optional

try:
    from .constants import validate_plan_path, PLANS_DIR
    from .atomic_write import atomic_write
except ImportError:
    # When imported directly via sys.path (not as a package)
    from constants import validate_plan_path, PLANS_DIR
    from atomic_write import atomic_write


# ---------------------------
# Constants
# ---------------------------

STATE_SCHEMA_VERSION = "1.0.0"

DEFAULT_REVIEW_ITERATIONS: Dict[str, int] = {
    "simple": 1,
    "medium": 1,
    "high": 2,
}


# ---------------------------
# Utilities
# ---------------------------

def eprint(*args: Any) -> None:
    """Print to stderr."""
    print(*args, file=sys.stderr)


# ---------------------------
# State File Management
# ---------------------------

def get_state_file_path(plan_path: str) -> Path:
    """Derive state file path from plan file path with security validation.

    The state file is stored adjacent to the plan file with a .state.json extension.
    This prevents state loss when session IDs change or temp files are cleaned up.

    Example: ~/.claude/plans/foo.md -> ~/.claude/plans/foo.state.json

    Raises:
        ValueError: If plan_path is invalid or insecure
    """
    # SECURITY: Validate path before any operations
    validated_path = validate_plan_path(plan_path)

    # State file is always adjacent to plan file
    return validated_path.with_suffix('.state.json')


def load_state(plan_path: str) -> Optional[Dict[str, Any]]:
    """Load state file with schema validation and migration."""
    try:
        state_file = get_state_file_path(plan_path)  # Validates path

        if not state_file.exists():
            return None

        state = json.loads(state_file.read_text(encoding="utf-8"))

        # Handle schema version (backward compatible)
        schema_version = state.get("schema_version")

        if schema_version is None:
            # Existing state files without version - auto-migrate
            state["schema_version"] = STATE_SCHEMA_VERSION
            eprint(f"[state] Migrated state file to schema v{STATE_SCHEMA_VERSION}")
        elif schema_version != STATE_SCHEMA_VERSION:
            eprint(f"[state] WARNING: Schema mismatch (expected {STATE_SCHEMA_VERSION}, got {schema_version})")
            # For now, accept with warning - add migration logic here if schema changes

        return state

    except ValueError as e:
        eprint(f"[state] SECURITY: Invalid plan path: {e}")
        return None
    except Exception as e:
        eprint(f"[state] ERROR: Failed to load state: {e}")
        return None


def save_state(plan_path: str, state: Dict[str, Any]) -> bool:
    """Save state file with schema version and validation.

    Returns True on success, False on failure.
    """
    try:
        state_file = get_state_file_path(plan_path)  # Validates path

        state_with_version = {
            "schema_version": STATE_SCHEMA_VERSION,
            **state
        }

        # Use atomic write
        success, error = atomic_write(
            state_file,
            json.dumps(state_with_version, indent=2)
        )

        if not success:
            eprint(f"[state] Failed to save state: {error}")
            return False

        return True

    except ValueError as e:
        eprint(f"[state] SECURITY: Invalid plan path: {e}")
        return False
    except Exception as e:
        eprint(f"[state] ERROR: {e}")
        return False


def delete_state(plan_path: str) -> bool:
    """Delete state file after successful archive.

    Returns True if deleted or didn't exist, False on error.
    """
    try:
        state_file = get_state_file_path(plan_path)
        if state_file.exists():
            state_file.unlink()
            eprint(f"[state] Deleted state file: {state_file}")
        return True
    except ValueError as e:
        eprint(f"[state] SECURITY: Invalid plan path in delete: {e}")
        return False
    except Exception as e:
        eprint(f"[state] Warning: failed to delete state file: {e}")
        return False


# ---------------------------
# Iteration State Management
# ---------------------------

def get_iteration_state(
    state: Dict[str, Any],
    complexity: str,
    config: Optional[Dict[str, Any]] = None,
) -> Dict[str, Any]:
    """Get or initialize iteration state based on complexity.

    Args:
        state: The current plan state dict
        complexity: Plan complexity level (simple/medium/high)
        config: Optional config dict with reviewIterations settings

    Returns:
        Iteration dict with: current, max, complexity, history
    """
    if "iteration" in state:
        # Return existing iteration state
        return state["iteration"]

    # Initialize new iteration state
    review_iterations = DEFAULT_REVIEW_ITERATIONS.copy()
    if config:
        review_iterations.update(config.get("reviewIterations", {}))
    max_iterations = review_iterations.get(complexity, 1)

    return {
        "current": 1,
        "max": max_iterations,
        "complexity": complexity,
        "history": [],
    }


def update_iteration_state(
    state: Dict[str, Any],
    iteration: Dict[str, Any],
    plan_hash: str,
    verdict: str,
) -> Dict[str, Any]:
    """Record review result in iteration history and update state.

    Args:
        state: The current plan state dict
        iteration: The iteration state dict
        plan_hash: Hash of the current plan content
        verdict: Review verdict (pass/warn/fail)

    Returns:
        Updated state dict with iteration data
    """
    # Add this review to history
    iteration["history"].append({
        "hash": plan_hash,
        "verdict": verdict,
        "timestamp": datetime.now().isoformat(),
    })

    # Update state with iteration data
    state["iteration"] = iteration
    return state


def should_continue_iterating(
    iteration: Dict[str, Any],
    verdict: str,
    config: Optional[Dict[str, Any]] = None,
) -> bool:
    """Determine if more review iterations are needed.

    Args:
        iteration: The iteration state dict
        verdict: Current review verdict
        config: Optional config dict with earlyExitOnAllPass setting

    Returns:
        True if more iterations needed, False otherwise
    """
    current = iteration.get("current", 1)
    max_iter = iteration.get("max", 1)

    # At or past max iterations - no more iterations
    if current >= max_iter:
        eprint(f"[state] At max iterations ({current}/{max_iter}), no more iterations")
        return False

    # Check early exit on all pass
    early_exit = True
    if config:
        early_exit = config.get("earlyExitOnAllPass", True)
    if early_exit and verdict == "pass":
        eprint(f"[state] All reviewers passed and earlyExitOnAllPass=true, exiting early")
        return False

    # More iterations available and verdict is not pass (or early exit disabled)
    eprint(f"[state] Continuing to next iteration ({current + 1}/{max_iter}), verdict={verdict}")
    return True

File: packages/cli/src/templates/cc-native/_cc-native/lib-ts/state.ts
/**
 * Iteration state management for plan review cycles.
 * State files are stored adjacent to plan files (e.g., foo.md → foo.state.json).
 * See cc-native-plan-review-spec.md §4.7
 */

import * as fs from "node:fs";
import * as path from "node:path";

import { validatePlanPath } from "./constants.js";
import type { IterationState, IterationEntry } from "./types.js";
import { atomicWrite } from "../../_shared/lib-ts/base/atomic-write.js";
import { logInfo, logWarn, logError } from "../../_shared/lib-ts/base/logger.js";
import { nowIso } from "../../_shared/lib-ts/base/utils.js";

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const STATE_SCHEMA_VERSION = "1.0.0";

export const DEFAULT_REVIEW_ITERATIONS: Record<string, number> = {
  simple: 1,
  medium: 3,
  high: 5,
};

// ---------------------------------------------------------------------------
// State File Management
// ---------------------------------------------------------------------------

/**
 * Derive state file path from plan file path with security validation.
 * Example: ~/.claude/plans/foo.md → ~/.claude/plans/foo.state.json
 *
 * @throws Error if planPath is invalid or insecure
 */
export function getStateFilePath(planPath: string): string {
  const validated = validatePlanPath(planPath);
  const parsed = path.parse(validated);
  return path.join(parsed.dir, `${parsed.name}.state.json`);
}

/**
 * Load state file with schema validation and migration.
 */
export function loadState(planPath: string): Record<string, unknown> | null {
  try {
    const stateFile = getStateFilePath(planPath);

    if (!fs.existsSync(stateFile)) {
      return null;
    }

    const state = JSON.parse(
      fs.readFileSync(stateFile, "utf-8"),
    ) as Record<string, unknown>;

    // Handle schema version (backward compatible)
    const schemaVersion = state.schema_version as string | undefined;

    if (schemaVersion === undefined) {
      state.schema_version = STATE_SCHEMA_VERSION;
      logInfo(
        "state",
        `Migrated state file to schema v${STATE_SCHEMA_VERSION}`,
      );
    } else if (schemaVersion !== STATE_SCHEMA_VERSION) {
      logWarn(
        "state",
        `Schema mismatch (expected ${STATE_SCHEMA_VERSION}, got ${schemaVersion})`,
      );
    }

    return state;
  } catch (error: unknown) {
    if (error instanceof Error && error.message.includes("Invalid plan path")) {
      logError("state", `SECURITY: Invalid plan path: ${error}`);
    } else {
      logError("state", `Failed to load state: ${error}`);
    }
    return null;
  }
}

/**
 * Save state file with schema version and validation.
 * Returns true on success, false on failure.
 */
export function saveStateToPlan(
  planPath: string,
  state: Record<string, unknown>,
): boolean {
  try {
    const stateFile = getStateFilePath(planPath);

    const stateWithVersion = {
      schema_version: STATE_SCHEMA_VERSION,
      ...state,
    };

    const [success, error] = atomicWrite(
      stateFile,
      JSON.stringify(stateWithVersion, null, 2),
    );

    if (!success) {
      logError("state", `Failed to save state: ${error}`);
      return false;
    }

    return true;
  } catch (error: unknown) {
    if (error instanceof Error && error.message.includes("Invalid plan path")) {
      logError("state", `SECURITY: Invalid plan path: ${error}`);
    } else {
      logError("state", String(error));
    }
    return false;
  }
}

/**
 * Delete state file after successful archive.
 * Returns true if deleted or didn't exist, false on error.
 */
export function deleteState(planPath: string): boolean {
  try {
    const stateFile = getStateFilePath(planPath);
    if (fs.existsSync(stateFile)) {
      fs.unlinkSync(stateFile);
      logInfo("state", `Deleted state file: ${stateFile}`);
    }
    return true;
  } catch (error: unknown) {
    if (error instanceof Error && error.message.includes("Invalid plan path")) {
      logError("state", `SECURITY: Invalid plan path in delete: ${error}`);
      return false;
    }
    logWarn("state", `Failed to delete state file: ${error}`);
    return false;
  }
}

// ---------------------------------------------------------------------------
// Iteration State Management
// ---------------------------------------------------------------------------

/**
 * Get or initialize iteration state based on complexity.
 */
export function getIterationState(
  state: Record<string, unknown>,
  complexity: string,
  config?: Record<string, unknown>,
): IterationState {
  if (state.iteration) {
    return state.iteration as IterationState;
  }

  // Initialize new iteration state
  const reviewIterations = { ...DEFAULT_REVIEW_ITERATIONS };
  if (config) {
    const overrides = config.reviewIterations as
      | Record<string, number>
      | undefined;
    if (overrides) {
      Object.assign(reviewIterations, overrides);
    }
  }

  return {
    current: 1,
    max: reviewIterations[complexity] ?? 1,
    complexity,
    history: [],
    graduated: [],
    passStreaks: {},
    lastPlanHash: "",
    lastPlanPath: "",
    sessionId: "",
  };
}

/**
 * Record review result in iteration history and update state.
 */
export function updateIterationState(
  state: Record<string, unknown>,
  iteration: IterationState,
  planHash: string,
  verdict: string,
): Record<string, unknown> {
  const entry: IterationEntry = {
    hash: planHash,
    verdict,
    timestamp: nowIso(),
  };

  iteration.history.push(entry);
  state.iteration = iteration;
  return state;
}

/**
 * Determine if more review iterations are needed.
 */
export function shouldContinueIterating(
  iteration: IterationState,
  verdict: string,
  config?: Record<string, unknown>,
): boolean {
  const {current} = iteration;
  const maxIter = iteration.max;

  // At or past max iterations
  if (current >= maxIter) {
    logInfo(
      "state",
      `At max iterations (${current}/${maxIter}), no more iterations`,
    );
    return false;
  }

  // Check early exit on all pass
  let earlyExit = true;
  if (config) {
    earlyExit = (config.earlyExitOnAllPass as boolean) ?? true;
  }
  if (earlyExit && verdict === "pass") {
    logInfo(
      "state",
      "All reviewers passed and earlyExitOnAllPass=true, exiting early",
    );
    return false;
  }

  logInfo(
    "state",
    `Continuing to next iteration (${current + 1}/${maxIter}), verdict=${verdict}`,
  );
  return true;
}

// ---------------------------------------------------------------------------
// Review-level Iteration State I/O (stored in reviews dir)
// ---------------------------------------------------------------------------

/**
 * Load iteration state from the reviews directory.
 */
export function loadIterationState(reviewsDir: string): IterationState | null {
  const iterationFile = path.join(reviewsDir, "iteration.json");
  if (!fs.existsSync(iterationFile)) return null;
  try {
    return JSON.parse(fs.readFileSync(iterationFile, "utf-8")) as IterationState;
  } catch (error) {
    logError("state", `Failed to load iteration state: ${error}`);
    return null;
  }
}

/**
 * Save iteration state to the reviews directory.
 */
export function saveIterationState(reviewsDir: string, state: IterationState): boolean {
  const iterationFile = path.join(reviewsDir, "iteration.json");
  try {
    fs.mkdirSync(reviewsDir, { recursive: true });
    const toWrite = { ...state, schema_version: "1.0.0" };
    fs.writeFileSync(iterationFile, JSON.stringify(toWrite, null, 2), "utf-8");
    return true;
  } catch (error) {
    logError("state", `Failed to save iteration state: ${error}`);
    return false;
  }
}
